source/common/http/conn_manager_impl.cc

```diff
@@ -141,6 +142,24 @@ void ConnectionManagerImpl::initializeReadFilterCallbacks(Network::ReadFilterCal
       {stats_.named_.downstream_cx_rx_bytes_total_, stats_.named_.downstream_cx_rx_bytes_buffered_,
        stats_.named_.downstream_cx_tx_bytes_total_, stats_.named_.downstream_cx_tx_bytes_buffered_,
        nullptr, &stats_.named_.downstream_cx_delayed_close_timeout_});
+
+  // register callback for drain-close events
+  start_drain_cb_ =
+      drain_close_.addOnDrainCloseCb([this](std::chrono::milliseconds drain_delay) -> void {
+        // The "onDrainClose" callback may be called from another thread, so wrap our lambda
+        // body with `post()` to make sure it is executing on this thread.
+        read_callbacks_->connection().dispatcher().post([this, drain_delay]() -> void {
+          start_drain_timer_ =
+              read_callbacks_->connection().dispatcher().createTimer([this]() -> void {
```

Comment:
I think there are still a few possible issues:
- read_callbacks_ is read across threads without synchronization, so the execution of the drain close cb could race with the read_callback termination sequence and call to removeOnDrainCloseCb
- related to the previous issue, what guarantees that read_callbacks_ is not deleted between the time the callback is posted and when it executes?

I think that we need a mechanism to track drain callbacks per worker and a way to propagate drains from the process wide object to the tracker per worker.  Other possible alternatives may include use of weak pointers or similar mechanisms to determine if read_callbacks_ is still live when attempting to execute the associated callback.  Also, the dispatcher associated with the read_callbacks_ should be captured explicitly so there's no cross-thread read of the read_callback_ data members.

Reply:
@antoniovicente 
I agree with you that it would be best to manage things per-worker and I want to bake this abstraction into the thread-safe version of the callback manager so that it can be re-used safely (and just feels like a proper encapsulation of scope). 

I'm considering the following approach, but want to get your take on it. `ThreadSafeCallbackManager` would change the `add` function as:

```cpp
ThreadSafeCallbackHandlePtr add(Event::Dispatcher&, Callback); 
```

The `ThreadSafeCallbackHandlePtr` would be a `shared_ptr` and a `weak_ptr` would be retained along with the `Disptacher`. The `runCallbacksWith` function would look roughly the same from an API perspective as it does now but internally will use the `Dispatcher` to `post()` on the thread that registered the callbacks.  

__Deletion__ \
Since the returned callback-handle is now a `shared_ptr`, we don't have to perform any cleanup within the callback manager. Instead we can trim callbacks that have expired during normal operations of the class such as getting the size() or executing the callbacks. 

Locks will still be required when updating the central list, but since all add/remove operations would now be local to the class, I believe this would remove any sort of re-entrant locking behaviors.

Reply:
I've implemented the refactor that I discussed above so resolving this conversation for now for re-review.